core/translate/plan.rs

```diff
@@ -0,0 +1,469 @@
+use core::fmt;
+use std::{
+    fmt::{Display, Formatter},
+    rc::Rc,
+};
+
+use sqlite3_parser::ast;
+
+use crate::{function::AggFunc, schema::BTreeTable, util::normalize_ident, Result};
+
+pub struct Plan {
+    pub root_node: Operator,
+    pub referenced_tables: Vec<(Rc<BTreeTable>, String)>,
+}
+
+impl Display for Plan {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.root_node)
+    }
+}
+
+/**
+  An Operator is a Node in the query plan.
+  Operators form a tree structure, with each having zero or more children.
+  For example, a query like `SELECT t1.foo FROM t1 ORDER BY t1.foo LIMIT 1` would have the following structure:
+    Limit
+      Order
+        Project
+          Scan
+*/
+#[derive(Clone, Debug)]
+pub enum Operator {
+    Aggregate {
+        id: usize,
+        source: Box<Operator>,
+        aggregates: Vec<Aggregate>,
+    },
+    Filter {
+        id: usize,
+        source: Box<Operator>,
+        predicates: Vec<ast::Expr>,
+    },
+    SeekRowid {
+        id: usize,
+        table: Rc<BTreeTable>,
+        table_identifier: String,
+        rowid_predicate: ast::Expr,
+        predicates: Option<Vec<ast::Expr>>,
+    },
+    Limit {
+        id: usize,
+        source: Box<Operator>,
+        limit: usize,
+    },
+    Join {
+        id: usize,
+        left: Box<Operator>,
+        right: Box<Operator>,
+        predicates: Option<Vec<ast::Expr>>,
```

Comment:
There will be a lot of cloning while parsing, but parsing is not a hot path anyways. Something to note for the future.